### PR TITLE
hex-literal: add support for hex prefixes

### PR DESCRIPTION
**Pull request by DaniPopes**
Allows writing `hex!("0x1234")` by ignoring the initial `"0x"` prefix.
This only works by stripping the start of the literals, so something like `hex!("0x12 0x34")` or `hex!("   0x1234")` wouldn't work.
This would require a bit more effort to implement, so please let me know if you want this as well.

### CODE REVIEW

**Comment by newpavlov** [comment]
See #895. I don't think it's a good idea to support `0x` prefixes, since it can result in endianess-confusion. You should prefer using the [`crypto-bigint`](https://github.com/RustCrypto/crypto-bigint) crate when working with hex-encoded bigint values.

**Comment by DaniPopes** [comment]
Makes sense for bigints, but this is just a wrapper around byte arrays so there should be no confusion IMO: `hex!("010203")` always is `[0x01, 0x02, 0x03]`
I think there should be the option to support this either way

**Comment by newpavlov** [comment]
`0x`-prefix is commonly means big-endian order. When you write `0x01234567u32` it gets represented as `[0x67, 0x45, 0x23, 0x01]` on little-endian machines. So by encountering `hex!(0x01234567)` in code one may be quite surprised to find `[0x01, 0x23, 0x45, 0x67]` in memory.

**Comment by Pzixel** [comment]
@newpavlov can you please consider revisiting this decision? The real-world application in my case for example is copying a lot of literals from other sources and using them as literals in the probram. In all cases these literals have 0x prefix, so need to tinker them manually in order to make macro parse it correctly is source of errors which isn't ideal. While I understand that from library perspective it may not be obvious but real-world usage shows that library would benefit from providing one way or another to express this. Maybe as an extra macro but I think it's very important to have it here.

**Comment by tarcieri** [comment]
The macro does map the big endian representation of an integer to an array which has big endian byte ordering.

If we added support for hex literals with other types, I would expect it to use an architecture-specific byte ordering, the same way a `0x...` big endian literal maps to a host-specific integer.

**Comment by newpavlov** [comment]
@Pzixel 
`hex-literal` is less than 100 lines long. IIUC you work with fixed size values, so you can write an even simpler `const fn` to do decoding for your project. It can look something like this:
```rust
pub const fn decode_u256(val: &str) -> [u8; 32] {
    let (prefix, bytes) = val.as_bytes().split_at(2);
    match prefix {
        b"0x" => (),
        _ => panic!("bad prefix"),
    }
    
    let mut i = 0;
    let mut buf = [0u8; 32];
    let mut buf_pos = 0;
    while i < bytes.len() {
        let mut pos = 0;
        while let Some((byte, new_pos)) = next_byte(bytes, pos) {
            buf[buf_pos] = byte;
            buf_pos += 1;
            pos = new_pos;
        }
        i += 1;
    }
    if buf.len() != buf_pos {
        panic!("length mismatch");
    }
    buf
}
```

**Comment by Pzixel** [comment]
I'm affraid I'm not quite following why endianess is a subject to change here. From my perspective `0x` prefix just sets that the constant is written in hex, which is a bit exsessive for a library called `hed-literal` yet I don't see a harm in supporting it. Maybe I'm missing something?

> hex-literal is less than 100 lines long

While this true and I already used code from this PR I thought it's a good idea to try to clarify this for future library users so they maybe could have an easier time working with literals.

Proper bigint types are unfortunately lacking a const-time parsing (but they do have const conversion from arrays) which is a cornerstone here. And as I said I have implemented my own workarounds but they are a bit project-specific so I thought it's a good idea to have something in basic libs so everyone in community could benefit from it. But the main reason here is probably that I really don't understand how prefix implies ordering in any way

**Comment by tarcieri** [comment]
> Proper bigint types are unfortunately lacking a const-time parsing

`crypto-bigint` supports constant-time parsing

**Comment by yjhmelody** [comment]
Hi @newpavlov But rust support `0x` syntax for int. But you know that integer is always understood literally and has nothing to do with big or small endianness.

**Comment by newpavlov** [comment]
@yjhmelody 
You said it yourself `0x` is used for integers. Endianness determines how those integers will be represented in memory. As I wrote above, it could makes sense to support `0x` for [`crypto_bigint::Uint`](https://docs.rs/crypto-bigint/latest/crypto_bigint/struct.Uint.html#method.from_le_hex). But the `hex!` macro creates a byte array, which is most certainly not an integer.

**Comment by yjhmelody** [comment]
I think this is a syntax issue, and byte order is just the underlying representation, and for types such as slices and bytes, there is no byte order.

**Comment by newpavlov** [comment]
I do not see your point. `0x` is traditionally used to indicate integer literals in using hex encoding in BE order. `hex!` is about byte arrays, not integers. If you code works with big integers, you should use `crypto-bigint` and its parsing utilities, not `hex-literal`.

I think we have properly articulated our opinion and there are no new points being provided, so it's unlikely we will change it.

**Comment by yjhmelody** [comment]
Isn't it traditionally said that 0x is used to indicate that a value is hexadecimal? Why is it related to big endian order? can you give some links

**Comment by tarcieri** [comment]
@yjhmelody `0x` is used as a sigil for describing an *integer literal* as hexadecimal bytes in big endian order, which is then converted into an integer in host endianness.

The `hex!` macro returns a byte array, not an integer. The literal syntax there is closer to `[0xAA, 0xBB, 0xCC, 0xDD]`.

`hex!` values are often interpreted as integers, but the endianness varies algorithm-by-algorithm. Older algorithms tend to use big endian; newer algorithms have in many cases shifted to little endian.

If this syntax were supported and wrote `u32::from_le_bytes(hex!("0xAABBCCDD"))`, it would decode to the value `0xDDCCBBAA` instead.

**Comment by Pzixel** [comment]
I'm with @yjhmelody in this one, I'm not quite following. `0xAABBCCDD` should mean exactly `[0xAA, 0xBB, 0xCC, 0xDD]`. It's not an integer, it's an array, and user can decide on endianess when converting array to the value. The only thing macro should do is converting string into array properly. Bytes should come in the same order as they are in the string. 

if there are uses like `let x: i32 = hex!("0x11223344)` then I understand your position, but integers have native language support already and they don't need an extra crate while you can absultely legally write just `let x: i32 = 0x11223344`

**Comment by tarcieri** [comment]
> 0xAABBCCDD should mean exactly [0xAA, 0xBB, 0xCC, 0xDD]. It's not an integer, it's an array

rustc would disagree with you:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=990f2cd9beef7875302a7b9e1c261bbd

```rust
assert_eq!(0xAABBCCDD, [0xAA, 0xBB, 0xCC, 0xDD])
```

```
error[E0277]: can't compare `{integer}` with `[{integer}; 4]`
 --> src/main.rs:2:5
  |
2 |     assert_eq!(0xAABBCCDD, [0xAA, 0xBB, 0xCC, 0xDD])
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `{integer} == [{integer}; 4]`
  |
```

That said, the purpose of the `0x` prefix/sigil in an integer literal is to signify that follows is a base 16 integer.

But we already have the arguably even more clear `hex!` prefix on existing literals.

What does `hex!("0x...")` provide that `hex!("...")` does not? It's redundant.

**Comment by yjhmelody** [comment]
> `u32::from_le_bytes(hex!("0xAABBCCDD"))`, it would decode to the value 0xDDCCBBAA instead.
it's `from_le_bytes` determining the repr not the hex syntax.

**Comment by Pzixel** [comment]
> What does hex!("0x...") provide that hex!("...") does not? It's redundant.

It allows you to copy-paste existing hexadecimal value without modification from the other sources that provide them. For example, you can copy an `etherscan` address and use it in the program with no modifications. You might find this minor but this is actually a big deal. I was finding a lot of bugs because extra symbols got cut etc.

Its' redundant but it allows more easy integration with other existing places where it is used. 

> rustc would disagree with you:

I was saying about your macro that returns `[u8; _]`, it doesn't return an integer, so I'm not sure what this proves tbh. If we already return an array (and we do) we can only arrange bytes in this array, and since there is only one arrangement that makes sense I don't see how endianess comes into play here.

So tldr would be like: on the internet 0x just means that value is in hex. Often times we want to use such a value in our codebase. And we would like to avoid requirement to strip the prefix - because while in codebase hex! clearly means that underlying value is hex, on the internet it's just a number, so they always get shipped with it - and when copy-pasting it (because typing it manually is another source of error) you can accidentally introduce a bug, so "copy & go" approach would be nice even if it means some redundancy.

**Comment by tarcieri** [comment]
> It allows you to copy-paste existing hexadecimal value without modification from the other sources that provide them.

But you have to add the `hex!("` on the front and `")` on the end anyway. Some modification is always required.

> I was saying about your macro that returns [u8; _], it doesn't return an integer, so I'm not sure what this proves tbh.

Integer literals with `0x` return an integer, not a byte array. That's the point.

> it's from_le_bytes determining the repr not the hex syntax.

No `from_le_bytes` is typically required with an integer literal, because it returns an integer, not bytes.

This is just going around in circles, so I'm going to lock it.
